azure-keyvault/src/main/java/com/microsoft/azure/keyvault/messagesecurity/HttpMessageSecurity.java

```diff
@@ -0,0 +1,330 @@
+            this.clientSignatureKey = MessageSecurityHelper.JsonWebKeyFromString(_clientSignatureKeyString);
+        }
+        if (_serverSignatureKeyString != null && !_serverSignatureKeyString.equals("")){
+            this.serverSignatureKey = MessageSecurityHelper.JsonWebKeyFromString(_serverSignatureKeyString);
+        }
+        if (_serverEncryptionKeyString != null && !_serverEncryptionKeyString.equals("")){
+            this.serverEncryptionKey = MessageSecurityHelper.JsonWebKeyFromString(_serverEncryptionKeyString);
+        }
+
+        this.clientEncryptionKey = MessageSecurityHelper.GenerateJsonWebKey();
+    }
+
+    /**
+     * Constructor (tests only)
+     *
+     * @param _clientSecurityToken
+     *      pop or bearer authentication token.
+     * @param _clientEncryptionKeyString
+     *      string with client signing key (public + private parts) or null if not supported
+     * @param _clientSignatureKeyString
+     *      string with client signing key (public + private parts) or null if not supported
+     * @param _serverEncryptionKeyString
+     *      string with server encryption key (public only) or null if not supported
+     * @param _serverSignatureKeyString
+     *      string with server signing key (public only) or null if not supported
+     * @param testMode
+     *      true for test mode (uses 0 for timestamp)
+     */
+    public HttpMessageSecurity(String _clientSecurityToken, String _clientEncryptionKeyString, String _clientSignatureKeyString, String _serverEncryptionKeyString, String _serverSignatureKeyString, boolean testMode) throws IOException{
+        this(_clientSecurityToken, _clientSignatureKeyString, _serverEncryptionKeyString, _serverSignatureKeyString);
+        this.testMode = testMode;
+        if (_clientEncryptionKeyString != null && !_clientEncryptionKeyString.equals("")){
+            this.clientEncryptionKey = MessageSecurityHelper.JsonWebKeyFromString(_clientEncryptionKeyString);
+        }
+    }
+
+    /**
+     * Protects existing request. Replaces its body with encrypted version.
+     *
+     * @param request
+     *      existing request.
+     *
+     * @return
+     *      new request with encrypted body if supported or existing request.
+     */
+    public Request protectRequest(Request request) throws IOException {
+        try{
+            Request result = request.newBuilder().header(AUTHENTICATE, BEARER_TOKEP_REFIX + clientSecurityToken).build();
+
+            if(!supportsProtection()) {
+                return result;
+            }
+
+            Buffer buffer = new Buffer();
+            request.body().writeTo(buffer);
+            String currentbody = buffer.readUtf8();
+
+            JsonWebKey clientPublicEncryptionKey = MessageSecurityHelper.GetJwkWithPublicKeyOnly(clientEncryptionKey);
+
+            String payload = currentbody.replace("}", ",\"rek\":{\"jwk\":" + clientPublicEncryptionKey.toString() + "}}");
```

Comment:
>String payload = currentbody.replace("}", ",\"rek\":{\"jwk\":" + clientPublicEncryptionKey.toString() + "}}"); [](start = 12, length = 110)

This seems like a dangerous way to add this to the JSON body.  At the moment I believe this will work for the methods that support message encryption, however if we apply it to a more complex JSON body it looks like it will replace internal '}' characters with the rek string not just the last one.